File: javascript/sorting_algorithms/quickSort.js
/*
  Quick sort using first element as pivot
*/

//function pivot(arr, start=0, end=arr.length + 1) {
function pivot(arr, start, end) {
    let pivot = arr[start];
    let swapIndex = start;

    const swap = (arr, indx1, indx2) => {
        [arr[indx1],arr[indx2]] = [arr[indx2],arr[indx1]];
    }
    // Compare pivot to each element
    for (let i = start + 1; i < arr.length; i++) {
        if (pivot > arr[i]) {
            swapIndex++;
            swap(arr, swapIndex, i);
            //console.log(arr);
        }
    }
    //Swap pivot with element with swapIndex
    swap(arr, start, swapIndex);
    //console.log(arr, pivot);
    return swapIndex;
}

//console.log(pivot([4, 8, 2, 1, 5, 7, 6, 3]));

//function quickSort(arr, left=0, right = arr.length - 1) {
function quickSort(arr, left, right) {
    if (left < right) {
        let pivotIndex = pivot(arr, left, right);
        //left
        quickSort(arr, left, pivotIndex - 1);
        //right
        quickSort(arr, pivotIndex + 1, right);
    }
    return arr;
}

//let arr1 = [4, 8, 2, 1, 5, 7, 6, 3];
let arr1 = [10,9,8,7,6,5,4,3,2,1,0];
console.log(quickSort(arr1, 0, arr1.length - 1));

File: python/sorting_algorithms/quick_sort.py
def partition(arr, low, high):
    pivot = arr[low]
    start = low + 1
    end = high

    while True:
        while start <= end and arr[end] >= pivot:
            end -= 1
        while start <= end and arr[start] <= pivot:
            start += 1

        if start <= end:
            arr[start], arr[end] = arr[end], arr[start]
        else:
            break
    arr[low], arr[end] = arr[end], arr[low]
    return end

def quick_sort(arr, start, end):
    if start < end: 
        pivot_index = partition(arr, start, end)
        quick_sort(arr, start, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, end)
    return arr

#arr1 = [4, 8, 2, 1, 5, 7, 6, 3]
arr1 = [4, -8, 2, 1, 0.5, 7, 6, 3, 0.49]
#print(partition(arr1, 0, len(arr1) - 1))
print(quick_sort(arr1, 0, len(arr1) - 1))

#arr1 = [26, 40, 41, 14, 18, 78, 19, 99]
#print(quick_sort(arr1, 0, len(arr1) - 1))

arr1 = [i for i in range(10, -1, -1)]
print(quick_sort(arr1, 0, len(arr1) - 1))
